tests: fix flakiness while starting PostgreSQL and PGHoard webserver

kmichel-aiven opened the pull request:
# About this change - What it does

The tests were waiting a fixed amount of time for the web server thread
and for PostgreSQL to start. This was too optimistic. Fix it by using an
threading Event or polling the process startup instead.

# Why this way

Because flaky tests are unpleasant.

kmichel-aiven committed:
Fix flaky wait for webserver startup in tests

Waiting a fixed amount of time does not work well with threads and
unpredictable load, use an event instead.

kmichel-aiven committed:
Fix flaky wait for PostgreSQL startup in tests

Waiting a fixed amount of time does not work well with threads and
unpredictable load, poll the process instead.

alanfranz reviewed (dismissed):
The change looks OK to me, but this appears to be always failing on 3.11 (I tried restarting once already).

```FAILED test/test_walreceiver.py::TestWalReceiver::test_walreceiver[12-foobar] - AssertionError: Expected at least 1 xlog uploads, got None```

kmichel-aiven committed:
Bump max-parallel to the number of python versions

We were waiting for the first 4 versions to complete before
starting the tests for 3.11.

kmichel-aiven committed:
Try harder to force walreceiver to finalize WAL files

Before this modification, `test_pause_on_disk_full` failed
occasionally. Locally it was usually once or twice every
100 test runs. With this modification, it didn't fail once
after ~600 runs.

`test_walreceiver` also uses `switch_wal` and might benefit
from this change. This test failed in CI but didn't
fail locally. I was not able to reproduce and validate the
improvement.

kmichel-aiven committed:
Fix flaky test_pause_on_disk_full

The deleter thread was too fast for the pause to trigger,
force it to be slower until we detect a pause event.
